components/eamxx/src/diagnostics/aerocom_cld.cpp

```diff
@@ -0,0 +1,208 @@
+
+  using PF = scream::PhysicsFunctions<DefaultDevice>;
+
+  std::vector<int> level_vector(m_nlevs);
+  for(int i = 1; i < m_nlevs; ++i) {
+    level_vector[i] = m_topbot == "Top" ? i : m_nlevs - i;
+  }
+
+  const auto out = m_diagnostic_output.get_view<Real **>();
+
+  // zero out the outputs
+  Kokkos::deep_copy(out, 0.0);
+
+  // Get the input fields
+  const auto tmid = get_field_in("T_mid").get_view<const Real **>();
+  const auto pden = get_field_in("pseudo_density").get_view<const Real **>();
+  const auto pmid = get_field_in("p_mid").get_view<const Real **>();
+  const auto qv   = get_field_in("qv").get_view<const Real **>();
+  const auto qc   = get_field_in("qc").get_view<const Real **>();
+  const auto qi   = get_field_in("qi").get_view<const Real **>();
+  const auto rel  = get_field_in("eff_radius_qc").get_view<const Real **>();
+  const auto rei  = get_field_in("eff_radius_qi").get_view<const Real **>();
+  const auto cld  = get_field_in("cldfrac_tot").get_view<const Real **>();
+  const auto nc   = get_field_in("nc").get_view<const Real **>();
+
+  // Hack: phony field to store dz
+  auto dz_f = get_field_in("pseudo_density").clone("dz");
+  auto dz   = dz_f.get_view<Real **>();
+
+  // Get gravity acceleration constant from constants
+  using physconst = scream::physics::Constants<Real>;
+  // TODO: move tunable constant to namelist
+  constexpr auto q_threshold = 0.0;
+  // TODO: move tunable constant to namelist
+  constexpr auto cldfrac_tot_threshold = 0.001;
+
+  const auto num_levs = m_nlevs;
+  const auto policy   = ESU::get_default_team_policy(m_ncols, m_nlevs);
+  Kokkos::parallel_for(
+      "Compute " + name(), policy, KOKKOS_LAMBDA(const MT &team) {
+        const int icol = team.league_rank();
+        // Subview the inputs at icol
+        auto tmid_icol = ekat::subview(tmid, icol);
+        auto pden_icol = ekat::subview(pden, icol);
+        auto pmid_icol = ekat::subview(pmid, icol);
+        auto qv_icol   = ekat::subview(qv, icol);
+        auto qc_icol   = ekat::subview(qc, icol);
+        auto qi_icol   = ekat::subview(qi, icol);
+        auto rel_icol  = ekat::subview(rel, icol);
+        auto rei_icol  = ekat::subview(rei, icol);
+        auto cld_icol  = ekat::subview(cld, icol);
+        auto nc_icol   = ekat::subview(nc, icol);
+
+        // We need dz too
+        // TODO: deduce dz_icol type, etc. on the spot here w/o field cloning?
+        auto dz_icol = ekat::subview(dz, icol);
+        PF::calculate_dz(team, pden_icol, pmid_icol, tmid_icol, qv_icol,
+                         dz_icol);
+
+        // Initialize the 1D "clear fraction" as 1 (totally clear)
+        auto clr_icol = 1.0;
+
+        // Loop over all layers in serial (due to accumulative
+        // product), starting at 2 (second highest) layer because the
+        // highest is assumed to hav no clouds for cldtop, but starting
+        // at m_nlevs-1 for cldbot
+
+        for(int ilay : level_vector) {
```

Comment:
If I am reading the code correctly (various column sums), we could do all this in a `Kokkos::parallel_reduce()` instead of serially. It is a bit more complex, but could be a good kokkos learning experience (if you are interested).

Reply:
I see, I was missing the product in there. Let me take a closer look.

Reply:
Yeah, there isn't really an easy solution I see. No "prefix product" parallel_scan exists in Kokkos. This is fine.